skxray/correlation.py

```diff
@@ -0,0 +1,371 @@
+                level += 1
+
+                # Checking whether there is next level for processing
+                processing = level < num_levels
+
+    # ending time for the process
+    end_time = time.time()
+
+    logger.info("Processing time for {0} images took {1} seconds."
+                "".format(n, (end_time - start_time)))
+
+    # the normalization factor
+    if len(np.where(past_intensity_norm == 0)[0]) != 0:
+        g_max = np.where(past_intensity_norm == 0)[0][0]
+    else:
+        g_max = past_intensity_norm.shape[0]
+
+    # g2 is normalized G
+    g2 = (G[: g_max] / (past_intensity_norm[: g_max] *
+                        future_intensity_norm[: g_max]))
+
+    # Convert from num_levels, num_bufs to lag frames.
+    tot_channels, lag_steps = core.multi_tau_lags(num_levels,
+                                                  num_bufs)
+    lag_steps = lag_steps[:g_max]
+
+    return g2, lag_steps
+
+
+def _process(buf, G, past_intensity_norm, future_intensity_norm,
+             labels, num_bufs, num_pixels, img_per_level, level, buf_no):
+    """
+    Internal helper function. This modifies inputs in place.
+
+    This helper function calculates G, past_intensity_norm and
+    future_intensity_norm at each level, symmetric normalization is used.
+
+    Parameters
+    ----------
+    buf : array
+        image data array to use for correlation
+
+    G : array
+        matrix of auto-correlation function without
+        normalizations
+
+    past_intensity_norm : array
+        matrix of past intensity normalizations
+
+    future_intensity_norm : array
+        matrix of future intensity normalizations
+
+    labels : array
+        labels of the required region of interests(roi's)
+
+    num_bufs : int, even
+        number of buffers(channels)
+
+    num_pixels : array
+        number of pixels in certain roi's
+        roi's, dimensions are : [number of roi's]X1
+
+    img_per_level : array
+        to track how many images processed in each level
+
+    level : int
+        the current multi-tau level
+
+    buf_no : int
+        the current buffer number
+
+    Notes
+    -----
+    :math ::
+        G   = <I(t)I(t + delay)>
+
+    :math ::
+        past_intensity_norm = <I(t)>
+
+    :math ::
+        future_intensity_norm = <I(t + delay)>
+
+    """
+    img_per_level[level] += 1
+
+    # in multi-tau correlation other than first level all other levels
+    #  have to do the half of the correlation
+    if level == 0:
+        i_min = 0
+    else:
+        i_min = num_bufs//2
+
+    for i in range(i_min, min(img_per_level[level], num_bufs)):
+        t_index = level*num_bufs/2 + i
+
+        delay_no = (buf_no - i) % num_bufs
+
+        past_img = buf[level, delay_no]
+        future_img = buf[level, buf_no]
+
+        #  get the matrix of auto-correlation function without normalizations
+        tmp_binned = (np.bincount(labels,
+                                  weights=np.ravel(past_img*future_img))[1:])
```

Comment:
I don't think you need this ravel as the data is raveled on L175

```python
tmp_binned = np.bincount(labels, weights=past_img*future_img)[1:]
```

should work fine and is more readable.